tibber_aws/aws_logs.py

```diff
@@ -0,0 +1,149 @@
+import logging
+from dataclasses import dataclass
+from datetime import datetime as dt
+import json
+from .aws_base import AwsBase
+from typing import Dict, List
+from types_aiobotocore_logs.client import CloudWatchLogsClient
+
+logger = logging.getLogger("aws_logs")
+
+
+@dataclass
+class CloudWatchLogEvent:
+    """An AWS CloudWatch log event (a log line)"""
+    logStreamName: str
+    timestamp: int
+    message: str
+    ingestionTime: int
+    eventId: str
+
+    def __post_init__(self):
+        self.timestamp_dt = dt.fromtimestamp(self.timestamp / 1000)
+        self.ingestionTime_dt = dt.fromtimestamp(self.ingestionTime / 1000)
+        self.message = json.loads(self.message)
+
+
+class Logs(AwsBase):
+    """CloudWatch Logs client to query AWS logs from a log stream."""
+    _client: CloudWatchLogsClient
+
+    def __init__(self, region_name="eu-west-1") -> None:
+        super().__init__("logs", region_name)
+
+    async def _get_log_events_raw(
+        self, event: str, log_group: str, start_time: dt, end_time: dt, extra_filter: dict = None, **kwargs
+    ) -> dict:
+        """Make a call to `CloudWatchLogsClient.filter_log_events` with optional filtering"""
+        await self.init_client_if_required()
+
+        properties_to_match = {"event": event}
+        if extra_filter is not None:
+            properties_to_match.update(extra_filter)
+        pattern = " && ".join([f'( $.{k} = "{v}" )' for k, v in properties_to_match.items()])
```

Comment:
This will break / allows query injection (eg. a bad person could set `v` to something containing a `"`, which then lets them do something like `" ]] "" == "`, to trick the filter into returning all log messages. 

IMO I'd argue to just take a `filterPattern` here - that way you have a well-documented filter language, and you don't need to change this code if you ever wanted to use other types of predicates and operators?

Reply:
What's a filterPattern?

If you want all log events, why bother injecting and not ask the client directly for `get_log_events`? What's the damage you mean it can cause?

Reply:
Well, so we're adding this to a library now right, and so imagine a year from now this gets used in some automated log processing tool somewhere. 

It's not clear to the caller that they can't pass any string into the filter dict here - any strings with quotes in them need to be escaped by the caller. 

Most likely that'll just lead to breakage when, inevitably, some string comes along that has a quote in it. However, it's also a bit bad security posture - we should try to not have code that allows query injection like this.

Say in the future someone is writing, I dunno, like an "automated GDPR extract-your-data" tool, and they use this API to extract logs pertaining to a particular user account. If the user is somehow able to choose the value of any of the filter parameters, they could add a `"` to the value, and then would be able to modify the query as they like, extracting any logs, rather than just the ones filtered for them. 

In this case that seems pretty unlikely, but it's still good practice to ensure code we put in libraries - expecting others to use it - is free of injection paths

Reply:
How does it differ from the native one? Where you can make your filter string from anything?

Reply:
I understand that it can be implicitly breaking. But, how does adding a filter creation method make it less secure than the native method allowing for ANY entry?

Reply:
discussed on the side! All gooood

Reply:
@mskytt Can you teach me? Interested in jakes response :)